pkg/controller/constraint/constraint_controller.go

```diff
@@ -136,16 +143,24 @@ type ReconcileConstraint struct {
 // Reconcile reads that state of the cluster for a constraint object and makes changes based on the state read
 // and what is in the constraint.Spec
 func (r *ReconcileConstraint) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	enabled := r.cs.Enter()
-	defer r.cs.Exit()
-	if !enabled {
-		r.log.Info("ignoring request, constraint controller disabled", "request", request)
+	gvk, unpackedRequest, err := util.UnpackRequest(request)
+	if err != nil {
+		// Unrecoverable, do not retry.
+		// TODO(OREN) add metric
+		log.Error(err, "unpacking request", "request", request)
+		return reconcile.Result{}, nil
+	}
+	// Sanity - make sure it is a constraint resource.
+	if gvk.Group != constraintsGroup || gvk.Version != constraintsVersion {
+		// Unrecoverable, do not retry.
+		log.Error(err, "invalid constraint GroupVersion", "gvk", gvk)
 		return reconcile.Result{}, nil
 	}
+	// TODO(OREN) do we need to ask the watch manager if we are registered for this kind?
```

Comment:
Asking the watch manager seems superfluous, as the channel wouldn't have received the event w/o registration.

Reply:
A scenario I was considering was when events are in-flight in the various queues and are received after we have deregistered the watch. I think concretely for this controller it will be fine - de-registration happens after the parent template is removed from OPA, by which point either constraints should have been removed by the garbage collector or if their cache lags, a lookup in OPA should fail and we should drop the request.